cloud_functions/gcs_event_based_ingest/gcs_ocn_bq_ingest/main.py

```diff
@@ -0,0 +1,459 @@
+    if external_query_sql:
+        print("EXTERNAL QUERY")
+        external_query(gcs_client, bq_client, gsurl, external_query_sql,
+                       dest_table_ref)
+        return
+
+    print("LOAD_JOB")
+    load_batches(gcs_client, bq_client, gsurl, dest_table_ref)
+
+
+def handle_duplicate_notification(
+    bkt: storage.Bucket,
+    success_blob: storage.Blob,
+    gsurl: str
+):
+    """
+    Need to handle potential duplicate Pub/Sub notifications.
+    To achieve this we will drop an empty "claimed" file that indicates
+    an invocation of this cloud function has picked up the success file
+    with a certain creation timestamp. This will support republishing the
+    success file as a mechanism of re-running the ingestion while avoiding
+    duplicate ingestion due to multiple Pub/Sub messages for a success file
+    with the same creation time.
+    """
+    success_blob.reload()
+    success_created_unix_timestamp = success_blob.time_created.timestamp()
+
+    claim_blob: storage.Blob = bkt.blob(
+        f"_claimed_{success_created_unix_timestamp}")
+    try:
+        claim_blob.upload_from_string(
+            "",
+            if_generation_match=0)
+    except PreconditionFailed as err:
+        raise RuntimeError(
+            f"The prefix {gsurl} appears to already have been claimed for "
+            f"{gsurl}{SUCCESS_FILENAME} with created timestamp"
+            f"{success_created_unix_timestamp}."
+            "This means that another invocation of this cloud function has"
+            "claimed the ingestion of this batch."
+            "This may be due to a rare duplicate delivery of the Pub/Sub "
+            "storage notification.") from err
+
+
+def external_query(gcs_client, bq_client, gsurl, query, dest_table_ref):
+    """Load from query over external table from GCS.
+
+    This hinges on a SQL query defined in GCS at _config/bq_transform.sql and
+    an external table definition _config/external.json (otherwise will assume
+    parquet external table)
+    """
+    external_table_config = read_gcs_file_if_exists(
+        gcs_client, f"{gsurl}_config/external.json")
+    if external_table_config:
+        external_table_def = json.loads(external_table_config)
+    else:
+        print(f"Falling back to default parquet external table."
+              f" {gsurl}/_config/external.json not found.")
+        external_table_def = DEFAULT_EXTERNAL_TABLE_DEFINITION
+
+    external_table_def["sourceUris"] = flatten(
+        get_batches_for_prefix(gcs_client, gsurl))
+    external_config = bigquery.ExternalConfig.from_api_repr(external_table_def)
+    job_config = bigquery.QueryJobConfig(
+        table_definitions={"temp_ext": external_config}, use_legacy_sql=False)
+    # for some reason query string literal wrapped in b''
+    rendered_query = str(
+        str(query).format(dest_dataset=dest_table_ref.dataset_id,
+                          dest_table=dest_table_ref.table_id))[2:-1]
```

Comment:
query is wrapped in b'' because it's a bytestring. Instead of [2:-1] just call --> query.decode('UTF-8')

Reply:
TIL thank you.